New Feature: fft_aggregated

aloosley opened the pull request:
Hi all,

As mentioned in the [Spectral Centroid PR](https://github.com/blue-yonder/tsfresh/pull/314), I decided it would be cleaner to just restart the PR.  As requested, I have added a new combiner feature called fft_aggregated.  

I added tests for it, they pass.  There are other tests that I'm not sure are related to this new feature that are not passing for me.

Best.

MaxBenChrist commented on tsfresh/feature_extraction/feature_calculators.py:
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
+        'Attribute must be "centroid", "variance", "skew", "kurtosis"'
+
+
+    def get_moment(y, moment):
```
can you please add a docstring here?

aloosley replied:
Done

MaxBenChrist commented on tsfresh/feature_extraction/feature_calculators.py:
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
+    def get_centroid(y):
+        return get_moment(y, 1)
+
+    def get_variance(y):
```
can you please add a docstring here?

aloosley replied:
Done

MaxBenChrist commented on tsfresh/feature_extraction/feature_calculators.py:
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
+    def get_moment(y, moment):
+        return y.dot(np.arange(len(y)) ** moment) / y.sum()
+
+    def get_centroid(y):
```
can you please add a docstring here?

aloosley replied:
Done

MaxBenChrist commented on tsfresh/feature_extraction/feature_calculators.py:
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
+    def get_kurtosis(y):
+        return kurtosis(y)
+
+    calculation = dict(
```
I like the way how you calculate the attributes by help of the calculation dictionary

aloosley replied:
Thanks.

MaxBenChrist commented on tests/units/feature_extraction/test_feature_calculations.py:
```diff
@@ -473,6 +473,47 @@ def test_fft_coefficient(self):
+        x = range(10)
+        res = pd.Series(dict(fft_aggregated(x, param)))
+        six.assertCountEqual(self, list(res.index), expected_index)
+        self.assertAlmostEqual(res['aggtype_centroid'], 1.135, places=3)
```
This looks like you ran the method on the input and fixated the result (in this case `1.135`).

So, this unit test does not check if the calculation is right. Instead it checks if the value of the feature is not changing between tsfresh version.

I would prefer a result that is calculated by hand. Can you think of a border case, where we can calculate the feature by hand?

MaxBenChrist commented on tsfresh/feature_extraction/feature_calculators.py:
```diff
@@ -826,6 +826,53 @@ def complex_agg(x, agg):
+    fft_abs = abs(np.fft.rfft(x))
+
+    res = [calculation[config["aggtype"]](fft_abs) for config in param]
+    index = ['aggtype_{}'.format(config["aggtype"]) for config in param]
```
the error is related to the inferring of a feature calculators parameter from the column name.

To correct this, you need to add quotation marks around the string paramter, it should be 

```
index = ['aggtype_"{}"'.format(config["aggtype"]) for config in param]
```

aloosley replied:
Done.  I updated the tests accordingly.

MaxBenChrist replied:
this line was not touched, was it?

MaxBenChrist commented:
Great work @aloosley, we are close to merge this into master 👍

MaxBenChrist commented on tests/units/feature_extraction/test_feature_calculations.py:
```diff
@@ -473,6 +473,47 @@ def test_fft_coefficient(self):
+            {"aggtype": "skew"},
+            {"aggtype": "kurtosis"}
+        ]
+        expected_index = ['aggtype_centroid', 'aggtype_variance', 'aggtype_skew', 'aggtype_kurtosis']
```
the expected index needs to have quotation marks around the attributes

aloosley replied:
Ok, I will push now, please note the tests will fail because I'm in the middle of fixing the kurtosis and skew calculations (it was incorrect to use the pandas.Series skew/kurtosis functions that I had committed).

aloosley commented:
Please give some time for me to finish fixing the skew and kurtosis calculations and to add a few tests.

MaxBenChrist commented:
> Please give some time for me to finish fixing the skew and kurtosis calculations and to add a few tests.

Sure sure, take your time :)

aloosley commented:
Everything should be set now.  I also fixed kurtosis and skew calculations for the edge cases where the discrete calculations blow up and are no longer meaningful. For example, with a dirac delta, the 0/0 division for kurtosis should limit to 3 but gives nonsense in the real world case because of noise and rounding. I also added hand calculated test values for x=sin() and x=Normal() cases.

aloosley commented:
One final push to I just fixed up a few bad comments I had glossed over.

MaxBenChrist commented:
Wow. Great work @aloosley! 

Happy to merge this one!

aloosley commented:
Thanks for the back and forth with me.  That was great.

MaxBenChrist commented:
> Thanks for the back and forth with me. That was great.

Thanks. Looking forward to see more feature calculators from you ;)
